### ISSUE TITLE
With-abstraction fools the productivity checker

### ISSUE THREAD

**Issue by laMudri** [agda/agda#3604]
Following on from [Thorsten's post to the mailing list](https://lists.chalmers.se/pipermail/agda/2019/010902.html), in the following code, Agda mistakenly accepts `f` as productive. We can use this to make the type checker loop on `test`.

```agda
data Maybe (A : Set) : Set where
  just : A → Maybe A
  nothing : Maybe A

data Is-nothing {A} : Maybe A → Set where
  nothing : Is-nothing nothing

record ℕ∞ : Set where
  coinductive
  field
    prd∞ : Maybe ℕ∞

open ℕ∞

f :  ℕ∞ → ℕ∞
prd∞ (f m) with prd∞ m
prd∞ (f m) | just pm with f pm
prd∞ (f m) | just pm | x = prd∞ x
prd∞ (f m) | nothing = nothing

inf∞ : ℕ∞
prd∞ inf∞ = just inf∞

test : Is-nothing (prd∞ (f inf∞))
test = nothing
```

If this
```
prd∞ (f m) | just pm with f pm
prd∞ (f m) | just pm | x = prd∞ x
```
is replaced by this
```
prd∞ (f m) | just pm = prd∞ (f pm)
```
the productivity checker will notice that this definition is not productive. But it believes that the version of `f` with the `with` is productive, and upon trying to run it fails to produce a result.

Agda version 2.5.4.2

**Comment by jespercockx** [comment]
Manually writing out the desugaring of `with` doesn't fool the productivity checker, so I assume this is yet another problem with the with-inliner. 
```agda

f :  ℕ∞ → ℕ∞
prd∞ (f m) = f' m (prd∞ m)
  where
    f' : ℕ∞ → Maybe ℕ∞ → Maybe ℕ∞
    f' m nothing   = nothing
    f' m (just pm) = f'' m pm (f pm)
      where
        f'' : ℕ∞ → ℕ∞ → ℕ∞ → Maybe ℕ∞
        f'' m pm x = prd∞ x
```
My suggestion is to get rid of with-inlining altogether, as it seems to create way more problems than it solves.

**Comment by nad** [comment]
> Manually writing out the desugaring of `with` doesn't fool the productivity checker, so I assume this is yet another problem with the with-inliner.

Agda also complains if `--without-K` is turned on, in which case the inlining is apparently turned off.

I wouldn't mind if the inlining is removed.

**Comment by MatthewDaggitt** [comment]
Just a heads up that removing the with-inlining has broken my code in 5 different places. Not saying it should be reverted, but just a friendly note that it's definitely causing me and [others](https://lists.chalmers.se/pipermail/agda/2019/011473.html) some headaches.

**Comment by MatthewDaggitt** [comment]
Another update. Still in the process of wading through my 10k LoC [routing library](https://github.com/MatthewDaggitt/agda-routing) trying to fix the problems raised by this change. I haven't yet finished, but so far I have found 43 instances of functions that were previously accepted as terminating but no longer are. Of these, in 30 of them the recursive call has no dependencies on the outcome of the `with` so have been "relatively" easy to fix by lifting the recursive call to the top level `with` as suggested by @gallais.

Of the remaining 13, the recursive call depends on the outcome of the `with` and therefore this trick cannot be used. Instead I've had to manually desugar all the `with` syntax via the pattern in the first comment in #59. Out of the 13, 5 only contained a single `with` but the remaining 8 have had up to 4 inner `with`s which makes the unrolling pretty horrific.

For example I'm in the middle of unrolling the following:
```agda
𝓒ₜ⊆𝓡ₜ : ∀ t {i p} → path (σ t X i j) ≈ₚ p → i ∈ᵤ 𝓒 t → i ∈ᵤ 𝓡 t
𝓒ₜ⊆𝓡ₜ t {i} {invalid}  p[σᵗXᵢⱼ]≈∅  _ = 𝓡-∅ t i p[σᵗXᵢⱼ]≈∅
𝓒ₜ⊆𝓡ₜ t {i} {valid []} p[σᵗXᵢⱼ]≈[] _ = 𝓡-[] t i p[σᵗXᵢⱼ]≈[]
𝓒ₜ⊆𝓡ₜ t {i} {valid ((_ , k) ∷ p ∣ _ ∣ _)} p[σᵗXᵢⱼ]≈ik∷p (i∈Sₜ , ik∷p∈Fₜ)
  with 𝓕-alignment t i∈Sₜ p[σᵗXᵢⱼ]≈ik∷p
... | refl , σᵗXᵢⱼ≈AᵢₖσᵗXₖⱼ , p[σᵗXₖⱼ]≈p with 𝓒-path t p[σᵗXᵢⱼ]≈ik∷p (i∈Sₜ , ik∷p∈Fₜ)
...   | valid ([ _ , k∈Fₜ ]∷ p∈Fₜ) with 𝓒ₜ⊆𝓡ₜ t p[σᵗXₖⱼ]≈p k∈Fₜ
...     | k∈Rₜ with Allₑ-resp-≈ₚ k∈Rₜ p[σᵗXₖⱼ]≈p
...       | valid pˡ = Allₑ-resp-≈ₚ (valid (σᵗXᵢⱼ≈AᵢₖσᵗXₖⱼ ∷ pˡ)) (≈ₚ-sym p[σᵗXᵢⱼ]≈ik∷p)
```
and it now has four nested functions and is 21 lines long.

I must confess I've now changed my mind, and would advocate that this change be reverted until #59 is solved and we have a replacement. My reasons are:
 - I'm already the second person reporting issues with this. I'm not sure what percentage of the Agda community tries out the release candidate, but it seems unlikely it's the majority. I'd imagine that an awful lot of people are going to be affected by this.
 - The required syntactic transformations to fix the problem are not obvious. I consider myself a relatively experienced Agda user (admittedly one with little knowledge of the internals) and I've struggled. I know that I wouldn't have had a clue when I started 2 years ago.
- The original bug which would be re-introduced if this change was reverted causes problems with invalid programs, where as the solution is causing problems with valid programs. Obviously I'm biased, but if I had to choose which problems to have I'd choose the former.

If it is not reverted, then at the very minimum the CHANGELOG entry needs to be greatly expanded with a detailed explanation of the different transformations necessary to work around the termination failure.

**Comment by MatthewDaggitt** [comment]
Hmm hang on I've just reread the mailing list thread. Unlike the top comment in this issue which said the type checker hangs, in the mailing list Thorsten shows an example which is incorrectly accepted as productive. If that is indeed the case, it does somewhat invalidate my third point.

**Comment by nad** [comment]
> For example I'm in the middle of unrolling the following:

Is that code structurally recursive?

**Comment by MatthewDaggitt** [comment]
> Is that code structurally recursive?

Yup, it's structurally recursive on the implicit parameter `p` which decreases from `valid ((_ , k) ∷ p ∣ _ ∣ _)` to `valid p` in the recursive call.

**Comment by gallais** [comment]
Have you considered performing a partially applied recursive call at the
top-level? i.e. `with valid {p}`?

**Comment by nad** [comment]
I guess you mean something like `𝓒ₜ⊆𝓡ₜ t {p = valid p}`.

**Comment by nad** [comment]
This is another trick that could be documented in the changelog.

**Comment by laMudri** [comment]
> Unlike the top comment in this issue which said the type checker hangs, in the mailing list Thorsten shows an example which is incorrectly accepted as productive.

Upon rereading, my description here is unclear. I'll edit that now.

**Comment by MatthewDaggitt** [comment]
> Have you considered performing a partially applied recursive call at the
top-level? i.e. with valid {p}?

Ah that's super sneaky :open_mouth: That should work in this case, I'll try it out tomorrow. Of course that only works in this case the structurally decreasing argument comes before the argument that depends on the with. If it's the other way round then I don't see how it would. Not sure off the top of my head if I have any of those cases....

**Comment by MatthewDaggitt** [comment]
> Have you considered performing a partially applied recursive call at the
top-level? i.e. with valid {p}?

Just attempted to implement this solution in the new definition below (@gallais I think this what you meant?):
```agda
𝓒ₜ⊆𝓡ₜ : ∀ t {i p} → path (σ t X i j) ≈ₚ p → i ∈ᵤ 𝓒 t → i ∈ᵤ 𝓡 t
𝓒ₜ⊆𝓡ₜ t {i} {invalid}  p[σᵗXᵢⱼ]≈∅  _ = 𝓡-∅ t i p[σᵗXᵢⱼ]≈∅
𝓒ₜ⊆𝓡ₜ t {i} {valid []} p[σᵗXᵢⱼ]≈[] _ = 𝓡-[] t i p[σᵗXᵢⱼ]≈[]
𝓒ₜ⊆𝓡ₜ t {i} {valid ((_ , k) ∷ p ∣ _ ∣ _)} p[σᵗXᵢⱼ]≈ik∷p (i∈Sₜ , ik∷p∈Fₜ)
  with 𝓒ₜ⊆𝓡ₜ t {i} {valid p} | 𝓕-alignment t i∈Sₜ p[σᵗXᵢⱼ]≈ik∷p
... | rec | refl , σᵗXᵢⱼ≈AᵢₖσᵗXₖⱼ , p[σᵗXₖⱼ]≈p with 𝓒-path t p[σᵗXᵢⱼ]≈ik∷p (i∈Sₜ , ik∷p∈Fₜ)
...   | valid ([ _ , k∈Fₜ ]∷ p∈Fₜ) with rec p[σᵗXₖⱼ]≈p k∈Fₜ
...     | k∈Rₜ with Allₑ-resp-≈ₚ k∈Rₜ p[σᵗXₖⱼ]≈p
...       | valid pˡ = Allₑ-resp-≈ₚ (valid (σᵗXᵢⱼ≈AᵢₖσᵗXₖⱼ ∷ pˡ)) (≈ₚ-sym p[σᵗXᵢⱼ]≈ik∷p)
```
Firstly there is the problem that I was wrong: the recursive call **does** depend on the result of the `with`s. Namely it depends on the pattern matching of `refl` in the first `with` statement. Therefore this solution isn't applicable in this particular case.

Secondly, I tried applying the same approach to a very similar structured proof:
```agda
𝓒-path : ∀ t {i p} → path (σ t X i j) ≈ₚ p → i ∈ᵤ 𝓒 t → Allᵥ (𝓒 t) p
𝓒-path t {i} {invalid}  _ _ = invalid
𝓒-path t {i} {valid []} _ _ = valid []
𝓒-path t {i} {valid ((_ , k) ∷ p ∣ _ ∣ _)} p[σᵗXᵢⱼ]≈ik∷p i∈𝓒ₜ@(i∈𝓕ₜ , ik∷p∈𝓕ₜ)
  with 𝓒-path t {i} {valid p} | 𝓕-alignment t i∈𝓕ₜ p[σᵗXᵢⱼ]≈ik∷p
... | rec | refl , _ , p[σᵗXₖⱼ]≈p with Allᵥ-resp-≈ₚ ik∷p∈𝓕ₜ p[σᵗXᵢⱼ]≈ik∷p
...   | (valid ([ _ , k∈𝓕ₜ ]∷ p∈𝓕ₜ)) with Allᵥ-resp-≈ₚ (valid p∈𝓕ₜ) (≈ₚ-sym p[σᵗXₖⱼ]≈p)
...     | k∈𝓒ₜ with rec p[σᵗXₖⱼ]≈p (k∈𝓕ₜ , k∈𝓒ₜ)
...       | valid p∈𝓒ₜ = valid ([ i∈𝓒ₜ , (k∈𝓕ₜ , k∈𝓒ₜ) ]∷ p∈𝓒ₜ)
```
and for some reason in this case Agda doesn't even accept the partial application `𝓒-path t {i} {valid p}` as terminating even when it has been lifted to the top `with` statement. Haven't checked whether this is a regression or whether it was present in 2.6.0.1

So in conclusion it doesn't look like that trick helps much for a couple of reasons...

**Comment by jespercockx** [comment]
> I must confess I've now changed my mind, and would advocate that this change be reverted until #59 is solved and we have a replacement.

The problem is that it is not a simple decision of reverting or not, since https://github.com/agda/agda/commit/87a6223ac3fb473e7b01ee9e0118184ea6725823 removed about 300 lines of code that were in dire need of maintenance. So at the very least, someone would have to take ownership of this piece of code and bring it up-to-date with the current version of Agda (I'm not that person, sorry).

I think it is a good idea to document in the changelog what techniques can work around the problem, but perhaps one of the other people in this tread have more hands-on experience and are in a better position to write such advice.

**Comment by nad** [comment]
My impression is that this feature was buggy and not well understood. Code that relies on it may well be incorrect, as noted above. For that reason I don't think we should keep the feature, even if the result is that some code breaks.

**Comment by jespercockx** [comment]
Are there any volunteers to improve the changelog entry with advice on how to migrate to the new version?

**Comment by UlfNorell** [comment]
I'll try to write something.

**Comment by UlfNorell** [comment]
@MatthewDaggitt abstracting the recursive call does work in your examples. You have the wrong value for `i` in your attempt. Abstracting over `𝓒ₜ⊆𝓡ₜ t {k} {valid p}` and `𝓒-path t {k} {valid p}` works for me.

**Comment by UlfNorell** [comment]
Guide in #4459.

**Comment by MatthewDaggitt** [comment]
> @MatthewDaggitt abstracting the recursive call does work in your examples. You have the wrong value for i in your attempt. Abstracting over 𝓒ₜ⊆𝓡ₜ t {k} {valid p} and 𝓒-path t {k} {valid p} works for me.

Oh (doh!). Thank you very much, that's saved me a lot of pain. Apologies for not spotting that myself.
